src/types/timestamp.rs

```diff
@@ -825,27 +769,49 @@ impl IonOrd for Timestamp {
 }
 
 /// A Builder object for incrementally configuring and finally instantiating a [Timestamp].
-/// For the time being, this type is not publicly visible. Users are expected to use any of the
-/// [TimeUnitSetter] implementations that wrap it. These wrappers expose only those methods which
-/// can result in a valid Timestamp. For example, it is not possible to set the `day` field
-/// without first setting the `year` and `month` fields.
+/// This builder uses the type-state pattern to expose only those methods which can result in a
+/// valid Timestamp. For example, it is not possible to set the `day` field without first setting
+/// the `year` and `month` fields.
 // See the unit tests for usage examples.
-#[derive(Debug, Clone, Default)]
-struct TimestampBuilder {
+#[derive(Debug, Clone)]
+pub struct TimestampBuilder<T> {
+    _state: PhantomData<T>,
     fields_are_utc: bool,
     precision: Precision,
     offset: Option<i32>,
-    year: u16,
-    month: Option<u8>,
-    day: Option<u8>,
-    hour: Option<u8>,
-    minute: Option<u8>,
-    second: Option<u8>,
+    year: u32,
+    month: Option<u32>,
+    day: Option<u32>,
+    hour: Option<u32>,
+    minute: Option<u32>,
+    second: Option<u32>,
```

Comment:
🗺️ As I was working on this change, I realized that there was an overflow bug. E.g.:

```rust
let ts1 = TimestampBuilder::new().with_ymd(1, 256, 256).build();
let ts2 = TimestampBuilder::new().with_ymd(1, 1, 1).build();
assert!(ts1.is_err()); // FAIL!
assert_ne!(ts1, ts2); // Also FAIL!
```

So, I changed all of the builder fields to `u32`. It made the most sense to me, especially since we were converting most of them back to `u32` when building the `Timestamp`. However, we may want to consider changing back to smaller integer types (without re-introducing the overflow bug, of course).

Reply:
The initial motivation for using `Option<u8>` instead of `Option<u32>` was to keep the memory layout from bloating. An `Option<u32>` often takes 8 bytes to represent despite only having 5 bytes of data due to alignment.

```rust
struct Foo {
    month: Option<u8>,
    day: Option<u8>,
    hour: Option<u8>,
    minute: Option<u8>,
    second: Option<u8>,
    nanoseconds: Option<u8>,
}

struct BigFoo {
    month: Option<u32>,
    day: Option<u32>,
    hour: Option<u32>,
    minute: Option<u32>,
    second: Option<u32>,
    nanoseconds: Option<u32>,
}

pub fn main() {
    println!("sizeof Foo: {}", std::mem::size_of::<Foo>());
    println!("sizeof BigFoo: {}", std::mem::size_of::<BigFoo>());
}

// ==== OUTPUT ====

sizeof Foo: 12
sizeof Foo: 48
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d64c6f640066aed3cef75ba05c71c1f1))

That said, it looks like even with the other fields added we're still looking at <100 bytes in all and I don't foresee people storing thousands of `TimestampBuilder`s, so it's probably fine. 👍

Reply:
We could also solve the overflow issue by shrinking the argument sizes. (E.g. `with_month(month: u32)` becomes `with_month(month: u8)`.

Since timestamps always represent a specific point in time, we could forgo the `Option` and set each field to the point in time it represents if the precision is less than that unit of time. For example, if the precision is "day", then it's not a big deal if the hour, minute, and second are all set to `0` instead of `None`.

Here's another [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b8c83aaa84919e107d4f0a5f50805df). (I've updated `Foo` to use `u32` for `nanoseconds` because we do actually need it there.) The output is:

```
sizeof Foo: 20
sizeof BigFoo: 48
sizeof MediumFoo: 24
sizeof LittleFoo: 12
```

What do you think? Is there any reason not to shrink the function argument sizes like `with_ymd(year: u16, month: u8, day: u8)`?

Reply:
IIRC, I used `u32` everywhere because other time-centric libraries like `chrono` did that. My sense is that it's kind of rare for APIs to use the most narrow int for a type, even though that's counter-intuitive. Maybe because it forces users to do verbose casting/conversion?